### ISSUE TITLE
Suggestion: Convolutional-neural-nets?

### ISSUE THREAD

**Issue by dobkeratops** [parallella/pal#224]
Somewhere between a fully fledged neural-net library, and the existing convolution functions in pal/image - would any of these elements be a good fit for the Pal library ?

imagine the following functions

3D x 3D -> 2D convolution, with clamped output (supply a minimum, e.g. zero,-1, -FLT_MAX ?), and optional ReLU pooling (N=1,2,3.. ? N=1 for no pooling); 

This would be a big chunk of the basic layer-evaluation of the deep-learning image recognition algorithms. Invoke multiple 3d convolutions for a 3D result. It would be important to include the clamp & ReLU since this would avoid significant memory traversals. You could provide a wrapper for a 3d x 3d->2d convolution without those steps that just calls it with (min=0, pooling=1)

An input could be (width x height x channels) - image-planes -  or a true 3D image.

Then imagine functions for training such a thing. (backpropogation and accumulating error deltas throught the weights).

I think a single step like that would go a long way to leveraging the epiphany hardware; you'd have a lot of data-reuse, perhaps uploading an entire 3D filter across multiple cores, then streaming an image through it .

This would be a stepping stone to a full neural net library which could implement pipelines between net layers. Getting some capability in the Pal library might make the epiphany chip more appealing to neural-net/Deep-learning researchers.

**Comment by jar** [comment]
I'm interested.  I think it would be good to first have a test case or proof-of-concept for the Epiphany.  Given the limitations of the core, I would like to see a proof-of-concept parallel implementation that can efficiently handle the inter-core communication to do something useful (even though all PAL functions are currently serial).

**Comment by dobkeratops** [comment]
there's no doubt in my mind it could do it - anything a cache can do, intercore-communication can - you just sometimes need to manually discriminate what is kept. (in the case of convolutions, I think most of the time you'd want to assume the kernel was kept local, but for completeness one might have to also have the option to stream the kernels? ) 

Another question is how far to go with intercore communication: a single function to do *multiple passes* would give you the opportunity to implement a whole pipeline on chip; however the API's would begin to explode in complexity; I got the impression PAL wants to be quite generic & straightforward. (you might need further details like describing streams of data)

e.g:-
<code>
    struct Convolution4dDesc {
        float* kernel;
        int kernel_size[4];/*4d tensor; supply trailing 1's for 2d or 3d..*/
        float* bias;/* for each slice of 4th axis*/ 
        float min; float max; /* output values clamped to range*/
        int xy_pooling_reduction_factor; /* optional reduction in 1st 2 axes, or 1 for no effect */
    }

    /* apply a series of convolutions (may use on chip communication between passes) */

    /* 3d data with a 4D kernel; any dimension being '1' can be skipped, 
        e.g. (32x32x64x1)image with 32x32x64x48 kernel
        produces an output that can be treated as  32x32x48 x1 or 32x32x1x48 for the next.*/

    void multi_conv4d(
       const float* input, int input_size[4] /*4d tensor,e.g. start with width x height x 3channels x 1 */
       float* output, /* output size is computed from convolutions & reductions*/
       const Convolution4dDesc* conv,int num_conv,
       float* workspace, /* buffer for temporary layer eval, contents is undefined after,can be null for 1layer */
    );

    void calc_conv4d_size(const Convolution4dDesc* in, int num_conv, int output_size[4], int *workspace_size);

</code>

**Comment by jar** [comment]
If you're streaming the kernel weights from off-chip memory, it seems to me that it's no longer a very good choice to use the Epiphany.  Copying the weights to perform a single multiply-accumulate per weight is not an effective use of the architecture because it quickly becomes bandwidth-bound.  The most effective network would include preconfigured weights and inter-core communication while streaming the input.  An effective inter-core communication scheme must be accounted for pretty much any network size that performs something useful.

Currently PAL can address the computation part, but the communication here is just not as straight forward as moving edges on image processing.  The communication buffers may be asymmetric in size and do not necessarily translate to nearest neighbors (on the on-chip network).  It's complicated.

Still, I agree that the prospect of doing it on the Epiphany is interesting since it does allow you to do the entire NN without explicit synchronization to global memory between layers like on a GPU.

**Comment by dobkeratops** [comment]
perhaps this can be helped by a single call to apply multiple layers to a batch of images. (upload kernel, upload image 1, classify, upload image 2, ... ) - then the implementation is at liberty to load them and setup an on chip data flow; or by providing a call to upload weights, then apply. 
They're typically up to 60mb of weights? Its going to be a while before there's an epiphany chip capable of holding all that; but even then, I thought holding the intermediate layer values on chip would still be a win. I guess if they do actually make a 1024core chip with 128k per core that could do it amazingly well

**Comment by jar** [comment]
Yes, thinking toward the future, a larger chip should do quite well. For now, it would be beat to focus on toy codes and proofs of concept. This is architecture and algorithm research since it's not like performance or energy efficiency records will be smashed.

I think inverting the data loading so that it's working on a batch of images doesn't solve the bandwidth limitation. It still uses the data just once. I could be wrong or misunderstanding.

**Comment by dobkeratops** [comment]
>>"working on a batch of images doesn't solve the bandwidth limitation. "

no; with 16 cores, that would seem to be an inherent limitation of todays' machine. It would simply provide an API that *could* run it faster, or run it much faster on sufficiently small datasets.
